File: data_handling/dataLoader.js
// data_handling/dataLoader.js
const fs = require('fs');
const csv = require('csv-parser');
const path = require('path');
const config = require('../config/recommenderConfig');
const { safeParseList } = require('../utils/helpers');

/**
 * Loads and preprocesses CSV data.
 * Returns a Promise resolving with { customerMap, productMap }.
 */
function loadAndPreprocessData() {
    console.log("Starting data loading...");
    const customerData = [];
    const productData = [];

    const customerPromise = new Promise((resolve, reject) => {
        fs.createReadStream(path.resolve(__dirname, '..', config.CUSTOMER_DATA_FILE))
            .pipe(csv())
            .on('data', (row) => customerData.push(row))
            .on('end', () => {
                console.log(`Customer CSV (${config.CUSTOMER_DATA_FILE}) processed.`);
                resolve(customerData);
            })
            .on('error', (error) => {
                console.error(`Error reading customer CSV: ${error.message}`);
                reject(new Error(`Failed to read customer data: ${error.message}`));
            });
    });

    const productPromise = new Promise((resolve, reject) => {
         fs.createReadStream(path.resolve(__dirname, '..', config.PRODUCT_DATA_FILE))
            .pipe(csv())
            .on('data', (row) => productData.push(row))
            .on('end', () => {
                console.log(`Product CSV (${config.PRODUCT_DATA_FILE}) processed.`);
                resolve(productData);
            })
            .on('error', (error) => {
                 console.error(`Error reading product CSV: ${error.message}`);
                 reject(new Error(`Failed to read product data: ${error.message}`));
            });
    });

    // Wait for both files to be read
    return Promise.all([customerPromise, productPromise])
        .then(([rawCustomers, rawProducts]) => {
            console.log("Preprocessing data...");

            // Process Customers into a Map for fast lookup
            const customerMap = new Map();
            rawCustomers.forEach(cust => {
                const customerId = cust[config.CUST_ID_COL];
                if (!customerId) return; // Skip rows without ID

                // Preprocess list columns
                cust[config.CUST_BROWSING_COL] = safeParseList(cust[config.CUST_BROWSING_COL]);
                cust[config.CUST_PURCHASE_COL] = safeParseList(cust[config.CUST_PURCHASE_COL]);
                // Convert numeric potentially
                cust[config.CUST_AVG_ORDER_COL] = parseFloat(cust[config.CUST_AVG_ORDER_COL]) || 0;
                cust[config.CUST_AGE_COL] = parseInt(cust[config.CUST_AGE_COL], 10) || null;

                customerMap.set(customerId, cust);
            });

            // Process Products into a Map
            const productMap = new Map();
            rawProducts.forEach(prod => {
                const productId = prod[config.PROD_ID_COL];
                if (!productId) return;

                // Preprocess list column
                prod[config.PROD_SIMILAR_COL] = safeParseList(prod[config.PROD_SIMILAR_COL]);
                // Preprocess numeric columns
                prod[config.PROD_RATING_COL] = parseFloat(prod[config.PROD_RATING_COL]) || 0;
                prod[config.PROD_PRICE_COL] = parseFloat(prod[config.PROD_PRICE_COL]) || 0;
                // Add other numeric conversions as needed

                productMap.set(productId, prod);
            });

            console.log(`Data preprocessed: ${customerMap.size} customers, ${productMap.size} products.`);
            return { customerMap, productMap }; // Return maps
        });
}

module.exports = { loadAndPreprocessData };

File: app/services/data_loader.py
import csv
from pathlib import Path
from typing import Dict, List, Tuple

from app.core.config import Settings


def safe_parse_list(value: str) -> List[str]:
    if not value or not isinstance(value, str):
        return []
    text = value.strip()
    if text.startswith("[") and text.endswith("]"):
        try:
            inner = text[1:-1]
            return [
                item.strip().strip("\"'")
                for item in inner.split(",")
                if item.strip()
            ]
        except Exception:
            return []
    return []


def _load_csv(path: Path) -> List[Dict[str, str]]:
    with path.open("r", newline="", encoding="utf-8") as handle:
        return list(csv.DictReader(handle))


def load_and_preprocess_data(settings: Settings) -> Tuple[Dict[str, dict], Dict[str, dict]]:
    customer_path = Path(settings.customer_data_file)
    product_path = Path(settings.product_data_file)

    raw_customers = _load_csv(customer_path)
    raw_products = _load_csv(product_path)

    customer_map: Dict[str, dict] = {}
    for customer in raw_customers:
        customer_id = customer.get(settings.cust_id_col)
        if not customer_id:
            continue
        customer[settings.cust_browsing_col] = safe_parse_list(
            customer.get(settings.cust_browsing_col, "")
        )
        customer[settings.cust_purchase_col] = safe_parse_list(
            customer.get(settings.cust_purchase_col, "")
        )
        try:
            customer[settings.cust_avg_order_col] = float(
                customer.get(settings.cust_avg_order_col, 0) or 0
            )
        except ValueError:
            customer[settings.cust_avg_order_col] = 0
        try:
            customer[settings.cust_age_col] = int(customer.get(settings.cust_age_col) or 0)
        except ValueError:
            customer[settings.cust_age_col] = None
        customer_map[customer_id] = customer

    product_map: Dict[str, dict] = {}
    for product in raw_products:
        product_id = product.get(settings.prod_id_col)
        if not product_id:
            continue
        product[settings.prod_similar_col] = safe_parse_list(
            product.get(settings.prod_similar_col, "")
        )
        try:
            product[settings.prod_rating_col] = float(
                product.get(settings.prod_rating_col, 0) or 0
            )
        except ValueError:
            product[settings.prod_rating_col] = 0
        try:
            product[settings.prod_price_col] = float(product.get(settings.prod_price_col, 0) or 0)
        except ValueError:
            product[settings.prod_price_col] = 0
        product_map[product_id] = product

    return customer_map, product_map
